omegaconf/basecontainer.py

```diff
@@ -435,7 +435,11 @@ def wrap(key: Any, val: Any) -> Node:
                 target = self._get_node(key)
                 if target is None:
                     if is_structured_config(val):
-                        ref_type = OmegaConf.get_type(val)
+                        element_type = self._metadata.element_type
+                        if element_type is Any:
+                            ref_type = OmegaConf.get_type(val)
+                        else:
+                            ref_type = element_type
```

Comment:
Can you explain this change? I am not sure what this is doing.

Reply:
Wrapping a value in a dictconfig that has element type with the ref_type of the node as the element type ensures that node is validated correctly. This solved the test case `{"user": User(age=1, name="foo")}` for `test_assign_wrong_type_to_dict`